sql: add feature usage telemetry for CTEs and subqueries

knz opened the pull request:
First commit from #35678.

Informs https://github.com/cockroachlabs/registration/issues/203

This adds telemetry for CTEs and subqueries.

knz committed:
sql: add feature usage telemetry for CTEs

Release note (sql change): CockroachDB will now report uses of
CTEs (`WITH ...`) in telemetry, when telemetry is enabled, so as to
guide future product roadmapping.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35650)
<!-- Reviewable:end -->

knz committed:
sql: add telemetry for usage of subqueries

Release note (sql change): CockroachDB will now report uses of
subqueries when telemetry is enabled, to guide future optimization
projects.

knz committed:
sql: add telemetry for usage of correlated subqueries

Release note: None

bobvawter reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> to my eyes, though I'm not intimately familiar with the planning code.

> Reviewed 7 of 7 files at r1, 4 of 4 files at r2, 4 of 4 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35650#-:-L_qulo3EIa1vQeTyDmf:bx42g0h)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo one comment

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35650#-:-L_rI2yU8IqKnF-qY3N3:bghmtuy)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, and @rytaft)

___
*[pkg/sql/opt/optbuilder/scalar.go, line 524 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35650#-L_rHWyw2fMs0lOsrVhO:-L_rHWyw2fMs0lOsrVhP:b-78v5pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1b25d1396f85cc3170de03c130076cae9ab85dd/pkg/sql/opt/optbuilder/scalar.go#L524)):*
> ```Go
> 
> 	// Register the use of correlation to telemetry.
> 	telemetry.Inc(sqltelemetry.CorrelatedSubqueryUseCounter)
> ```

Can this happen more than once in the same query? Maybe check if `b.IsCorrelated` already.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35650)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @bobvawter, @jordanlewis, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/scalar.go, line 524 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35650#-L_rHWyw2fMs0lOsrVhO:-L_rWhJj1X-VdOmMRTrV:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1b25d1396f85cc3170de03c130076cae9ab85dd/pkg/sql/opt/optbuilder/scalar.go#L524)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Can this happen more than once in the same query? Maybe check if `b.IsCorrelated` already.
</blockquote></details>

Good idea. Done.


<!-- Sent from Reviewable.io -->

knz commented:
TFYRs!

bors r+

rytaft commented:
___
*[pkg/sql/opt/optbuilder/scalar.go, line 524 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35650#-L_rHWyw2fMs0lOsrVhO:-L_whJZUE8lOXp4x-tVk:bpnoo1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1b25d1396f85cc3170de03c130076cae9ab85dd/pkg/sql/opt/optbuilder/scalar.go#L524)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Good idea. Done.
</blockquote></details>

But it seems like other metrics might be incremented more than once per query (e.g., `SubqueryUseCounter`). Is there a reason this one should be different? Or perhaps the others should be changed to once per query as well? (Sorry for the late comment)


<!-- Sent from Reviewable.io -->

knz commented:
> But it seems like other metrics might be incremented more than once per query (e.g., SubqueryUseCounter). Is there a reason this one should be different? Or perhaps the others should be changed to once per query as well? (Sorry for the late comment)

No there's no real reason this one should be different.

1) it's not so clear _how_ to increment the others just once per query
2) I believe that it's not _useful_ -- these metrics exist to gauge feature usage by client/cluster, perhaps app_name (and then binary: yes/no), not per query. We use numeric counters to establish the boolean value (0 = not used; >0 = used), but that's merely an abuse of the machinery.

@awoods187 please see the above. The discussion is a reminder that we won't be able to extract meaningful information from the magnitude of a counter on a single cluster. This is true not only of the subquery stuff discussed here, but more generally of the entire SQL telemetry we've set up so far.
